source/h2_frames.c

```diff
@@ -0,0 +1,1618 @@
+                    mask = s_literal_no_forward_save_mask;
+                    prefix_size = 4;
+                    break;
+                default:
+                    /* Unreachable */
+                    AWS_FATAL_ASSERT(false);
+            }
+        }
+
+        /* Write the top bits to signal representation */
+        output->buffer[output->len] = mask;
+
+        /* Write the index if indexed, or 0 to signal literal name */
+        if (aws_hpack_encode_integer(index, prefix_size, output)) {
+            return AWS_OP_ERR;
+        }
+
+        /* Names and values must be copied to avoid modifying the original struct */
+        struct aws_byte_cursor scratch;
+
+        if (!found_value) {
+            /* If not an indexed header, check if the name needs to be written */
+            if (!index) {
+                scratch = field->header.name;
+                if (aws_hpack_encode_string(encoder->hpack, &scratch, encoder->use_huffman, output)) {
+                    return AWS_OP_ERR;
+                }
+                AWS_ASSERT(scratch.len == 0);
+            }
+
+            /* Value must be written if the field isn't pure indexed */
+            scratch = field->header.value;
+            if (aws_hpack_encode_string(encoder->hpack, &scratch, encoder->use_huffman, output)) {
+                return AWS_OP_ERR;
+            }
+            AWS_ASSERT(scratch.len == 0);
+        }
+
+        if (field->hpack_behavior == AWS_H2_HEADER_BEHAVIOR_SAVE) {
+            /* Save for next time */
+            aws_hpack_insert_header(encoder->hpack, &field->header);
+        }
+    }
+
+    return AWS_OP_SUCCESS;
+}
+int aws_h2_frame_header_block_decode(
+    struct aws_h2_frame_header_block *header_block,
+    struct aws_h2_frame_decoder *decoder) {
+    AWS_PRECONDITION(header_block);
+    AWS_PRECONDITION(decoder);
+
+    /* Don't need to call init, frames will have done that */
+
+    struct aws_h2_frame_header_field field;
+
+    while (decoder->payload.len) {
+        AWS_ZERO_STRUCT(field);
+
+        uint8_t first_byte = *decoder->payload.ptr;
+
+        if (first_byte & s_indexed_header_field_mask) {
+            /* Indexed header, fetch from hpack */
+            field.hpack_behavior = AWS_H2_HEADER_BEHAVIOR_SAVE;
+
+            uint64_t index = 0;
+            if (aws_hpack_decode_integer(&decoder->payload, 7, &index)) {
+                return aws_raise_error(AWS_H2_ERR_COMPRESSION_ERROR);
+            }
+
+            if (index > SIZE_MAX) {
+                return aws_raise_error(AWS_H2_ERR_COMPRESSION_ERROR);
+            }
+
+            const struct aws_http_header *header = aws_hpack_get_header(decoder->hpack, (size_t)index);
+            if (!header) {
+                return aws_raise_error(AWS_H2_ERR_COMPRESSION_ERROR);
+            }
+            field.header = *header;
+
+            if (aws_array_list_push_back(&header_block->header_fields, &field)) {
+                return AWS_OP_ERR;
```

Comment:
every other error in this function results in aws_raise_error(AWS_H2_ERR_COMPRESSION_ERROR);

Reply:
Looked at this one, and I don't think it's a compression error because there's nothing wrong with the data.